onadata/apps/logger/data_migration/tree.py

```diff
@@ -0,0 +1,93 @@
+from functools import partial
+
+
+def rev_getattr(param):
+    return lambda obj: getattr(obj, param)
+
+
+class Tree(object):
+    def __init__(self, label='root', parent=None, children=None):
+        self.label = label
+        self.parent = parent
+        self.children = []
+        list(map(self.add_child, (children or [])))
+
+    def __repr__(self):
+        return self.label
+
+    def add_child(self, node):
+        assert isinstance(node, Tree)
+        self.children.append(node)
+
+    def find_child_by_label(self, label):
+        return next(
+            [child for child in self.children if child.label == label], None
+        )
+
+    def get_child_attribs(self, attrib):
+        return map(rev_getattr(attrib), self.children)
+
+    @property
+    def is_leaf(self):
+        return self.children == []
+
+    def search_node_by_label(self, label):
+        return self._dfs_search_node(self, label)
+
+    @classmethod
+    def _dfs_search_node(cls, tree, label):
+        for child in tree.children:
+            if child.is_leaf and child.label == label:
+                return child
+            child_search = cls._dfs_search_node(child, label)
+            if child_search:
+                return child_search
+        return None
+
+    def extract_ancestors_labels(self):
+        labels = []
+        while self.parent is not None:
+            labels.append(self.parent.label)
+        return labels
```

Comment:
This looks like infinite loop if `self.parent` is not none

Reply:
Done